core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java

```diff
@@ -514,6 +544,185 @@ public XContentBuilder toInnerXContent(XContentBuilder builder, Params params) t
+        internalSearchHit.highlightFields(highlightFields);
+        internalSearchHit.sourceRef(parsedSource);
+        internalSearchHit.explanation(explanation);
+        internalSearchHit.setInnerHits(innerHits);
+        if (matchedQueries.size() > 0) {
+            internalSearchHit.matchedQueries(matchedQueries.toArray(new String[matchedQueries.size()]));
+        }
+        if (shardId != null || nodeId != null) {
+            internalSearchHit.shard(new SearchShardTarget(nodeId, shardId));
+        }
+        internalSearchHit.fields(fields);
+        return internalSearchHit;
+    }
+
+    private static Explanation parseExplanation(XContentParser parser) throws IOException {
+        XContentParser.Token token;
+        float value = 0.0f;
+        String description = null;
+        List<Explanation> details = new ArrayList<>();
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            ensureExpectedToken(XContentParser.Token.FIELD_NAME, token, () -> parser.getTokenLocation());
+            String currentFieldName = parser.currentName();
+            token = parser.nextToken();
+            if (Fields.VALUE.equals(currentFieldName)) {
+                value = parser.floatValue();
+            } else if (Fields.DESCRIPTION.equals(currentFieldName)) {
+                description = parser.textOrNull();
+            } else if (Fields.DETAILS.equals(currentFieldName)) {
+                ensureExpectedToken(XContentParser.Token.START_ARRAY, token, () -> parser.getTokenLocation());
+                while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                    details.add(parseExplanation(parser));
+                }
+            } else {
+                throwUnknownField(currentFieldName, parser.getTokenLocation());
+            }
+        }
+        return Explanation.match(value, description, details);
+    }
+
+    private static Pattern shardIdParsePattern = Pattern.compile("[\\[]([^\\]]*)[\\]][\\[]([^\\]]*)[\\]]");
+
+    /**
+     * Parses a ShardId instance from its toString() representation that is used in rest responses.
+     * The expected format is [indexName][shardId].
+     */
+    static ShardId parseShardId(String text, XContentLocation location) {
+        Matcher matcher = shardIdParsePattern.matcher(text);
+        if (matcher.matches()) {
+            // we loose information about index uuid when parsing from the REST response, so we can only use UNKNOWN_UUID here
+            return new ShardId(new Index(matcher.group(1), ClusterState.UNKNOWN_UUID), Integer.parseInt(matcher.group(2)));
+        } else {
+            String message = "Failed to parse \"_shard\": expecting field with format [[indexName][shardId]] but found [%s]";
+            throw new ParsingException(location, String.format(Locale.ROOT, message, text));
+        }
+    }
+
+    private static Object parseValue(XContentParser parser) throws IOException {
```

Comment:
I made this public and reuse it from GetField. Still wondering if this should go to another more common classe though, but I don't have a better ides right now.